src/Two/AbstractProvider.php

```diff
@@ -200,9 +200,9 @@ protected function hasInvalidState()
             return false;
         }
 
-        $session = $this->request->getSession();
+        $state = $this->request->getSession()->pull('state');
 
-        return ! ($this->request->input('state') === $session->get('state'));
+        return ! (strlen($state) > 0 && $this->request->input('state') === $state);
```

Comment:
The "pull" method on the session removed state after its retrieved. So you
can only attempt to guess any given state exactly one time. So, timing
attack wouldn't be possible here since you can only try once.

On Tuesday, August 4, 2015, Mark Fox <notifications@github.com> wrote:

> In src/Two/AbstractProvider.php
> <https://github.com/laravel/socialite/pull/93#discussion_r36254365>:
>
> >
> > -        return ! ($this->request->input('state') === $session->get('state'));
> > +        return ! (strlen($state) > 0 && $this->request->input('state') === $state);
>
> Sorry to swoop in, I found this PR via FriendsOfPhp
> <https://github.com/FriendsOfPHP/security-advisories/pull/78/files> and
> was just glancing around and wonder if the above comparison is susceptible
> to a timing attack <https://en.wikipedia.org/wiki/Timing_attack>? If an
> attacker can control $this->request->input('state') then strict string
> equality may be leaking information.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/socialite/pull/93/files#r36254365>.
>